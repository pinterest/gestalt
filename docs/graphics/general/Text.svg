<svg xmlns="http://www.w3.org/2000/svg" width="312" height="180" fill="none"><rect width="114" height="8" x="46" y="62" fill="#005C62" rx="4"/><rect width="166" height="8" x="46" y="94" fill="#005C62" rx="4"/><rect width="220" height="8" x="46" y="78" fill="#005C62" rx="4"/><rect width="185" height="8" x="46" y="110" fill="#005C62" rx="4"/></svg>