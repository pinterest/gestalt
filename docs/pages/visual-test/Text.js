// @flow strict
import { type Node, Fragment } from 'react';
import { Box, Text } from 'gestalt';

export default function TextSpec(): Node {
  return (
    <Fragment>
      <Box padding={1}>
        <Text size="sm">Text size small</Text>
      </Box>
      <Box padding={1}>
        <Text size="md">Text size medium</Text>
      </Box>
      <Box padding={1}>
        <Text size="lg">Text size large</Text>
      </Box>
    </Fragment>
  );
}
