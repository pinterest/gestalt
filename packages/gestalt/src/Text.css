/*
This file was generated by running

  $ /Users/chrislloyd/.nvm/versions/node/v9.2.1/bin/node /Users/chrislloyd/src/gestalt/scripts/text.js

*/

:root {
  --font-size-1: 12px;
  --font-size-2: 14px;
  --font-size-3: 16px;
  --font-size-4: 18px;
  --font-size-5: 21px;
}

.Text {
  composes: antialiased sansSerif letterSpacing from "./Typography.css";
}

.fontSize1 {
  font-size: var(--font-size-1);
}

.fontSize2 {
  font-size: var(--font-size-2);
}

.fontSize3 {
  font-size: var(--font-size-3);
}

.fontSize4 {
  font-size: var(--font-size-4);
}

.fontSize5 {
  font-size: var(--font-size-5);
}

@media (--sm) {
  .smFontSize1 {
    font-size: var(--font-size-1);
  }

  .smFontSize2 {
    font-size: var(--font-size-2);
  }

  .smFontSize3 {
    font-size: var(--font-size-3);
  }

  .smFontSize4 {
    font-size: var(--font-size-4);
  }

  .smFontSize5 {
    font-size: var(--font-size-5);
  }
}

@media (--md) {
  .mdFontSize1 {
    font-size: var(--font-size-1);
  }

  .mdFontSize2 {
    font-size: var(--font-size-2);
  }

  .mdFontSize3 {
    font-size: var(--font-size-3);
  }

  .mdFontSize4 {
    font-size: var(--font-size-4);
  }

  .mdFontSize5 {
    font-size: var(--font-size-5);
  }
}

@media (--lg) {
  .lgFontSize1 {
    font-size: var(--font-size-1);
  }

  .lgFontSize2 {
    font-size: var(--font-size-2);
  }

  .lgFontSize3 {
    font-size: var(--font-size-3);
  }

  .lgFontSize4 {
    font-size: var(--font-size-4);
  }

  .lgFontSize5 {
    font-size: var(--font-size-5);
  }
}
