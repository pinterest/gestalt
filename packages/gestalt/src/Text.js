// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import colors from './Colors.css';
import styles from './Text.css';
import typography from './Typography.css';

const SIZE_SCALE: { [size: ?string]: number } = {
  xs: 1,
  sm: 2,
  md: 3,
  lg: 4,
  xl: 5,
};

type Props = {|
  align?: 'left' | 'right' | 'center' | 'justify',
  bold?: boolean,
  children?: React.Node,
  color?:
    | 'green'
    | 'pine'
    | 'olive'
    | 'blue'
    | 'navy'
    | 'midnight'
    | 'purple'
    | 'orchid'
    | 'eggplant'
    | 'maroon'
    | 'watermelon'
    | 'orange'
    | 'darkGray'
    | 'gray'
    | 'lightGray'
    | 'red'
    | 'white',
  inline?: boolean,
  italic?: boolean,
  overflow?: 'normal' | 'breakWord',
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  smSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  mdSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  lgSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  leading?: 'tall' | 'short',
  truncate?: boolean,
  __dangerouslyIncreaseLineHeight?: boolean,
|};

export default function Text({
  align = 'left',
  bold = false,
  children,
  color = 'darkGray',
  inline = false,
  italic = false,
  overflow = 'breakWord',
  size = 'md',
  smSize,
  mdSize,
  lgSize,
  leading,
  truncate = false,
  __dangerouslyIncreaseLineHeight = false,
}: Props) {
  const scale = SIZE_SCALE[size];
  const smScale = SIZE_SCALE[smSize];
  const mdScale = SIZE_SCALE[mdSize];
  const lgScale = SIZE_SCALE[lgSize];

  const cs = cx(
    styles.Text,
    styles[`fontSize${scale}`],
    lgSize && styles[`lgFontSize${lgScale}`],
    mdSize && styles[`mdFontSize${mdScale}`],
    smSize && styles[`smFontSize${smScale}`],
    color === 'blue' && colors.blue,
    color === 'darkGray' && colors.darkGray,
    color === 'eggplant' && colors.eggplant,
    color === 'gray' && colors.gray,
    color === 'green' && colors.green,
    color === 'lightGray' && colors.lightGray,
    color === 'maroon' && colors.maroon,
    color === 'midnight' && colors.midnight,
    color === 'navy' && colors.navy,
    color === 'olive' && colors.olive,
    color === 'orange' && colors.orange,
    color === 'orchid' && colors.orchid,
    color === 'pine' && colors.pine,
    color === 'purple' && colors.purple,
    color === 'red' && colors.red,
    color === 'watermelon' && colors.watermelon,
    color === 'white' && colors.white,
    leading === 'short' && typography.leadingShort,
    (leading === 'tall' || __dangerouslyIncreaseLineHeight) &&
      typography.leadingTall,
    align === 'center' && typography.alignCenter,
    align === 'justify' && typography.alignJustify,
    align === 'left' && typography.alignLeft,
    align === 'right' && typography.alignRight,
    overflow === 'breakWord' && typography.breakWord,
    italic && typography.fontStyleItalic,
    !italic && typography.fontStyleNormal,
    bold && typography.fontWeightBold,
    !bold && typography.fontWeightNormal,
    truncate && typography.truncate
  );
  const Tag = inline ? 'span' : 'div';

  return (
    <Tag
      className={cs}
      {...(truncate && typeof children === 'string'
        ? { title: children }
        : null)}
    >
      {children}
    </Tag>
  );
}

Text.propTypes = {
  __dangerouslyIncreaseLineHeight: PropTypes.bool,
  align: PropTypes.oneOf(['left', 'right', 'center', 'justify']),
  bold: PropTypes.bool,
  children: PropTypes.node,
  color: PropTypes.oneOf([
    'green',
    'pine',
    'olive',
    'blue',
    'navy',
    'midnight',
    'purple',
    'orchid',
    'eggplant',
    'maroon',
    'watermelon',
    'orange',
    'darkGray',
    'gray',
    'lightGray',
    'red',
    'white',
  ]),
  inline: PropTypes.bool,
  italic: PropTypes.bool,
  leading: PropTypes.oneOf(['tall', 'short']),
  lgSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  mdSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  overflow: PropTypes.oneOf(['normal', 'breakWord']),
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  smSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  truncate: PropTypes.bool,
};
