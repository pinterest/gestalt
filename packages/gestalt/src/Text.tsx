import { forwardRef, ReactElement, ReactNode } from 'react';
import cx from 'classnames';
import styles from './Text.css';
import { semanticColors } from './textTypes';
import typography from './Typography.css';

function isNotNullish(val?: number | null): boolean {
  return val !== null && val !== undefined;
}

type As = 'span' | 'div';
type Overflow = 'normal' | 'breakAll' | 'breakWord' | 'noWrap';
type Size = '100' | '200' | '300' | '400' | '500' | '600';

type Props = {
  /**
   * `"start"` and `"end"` should be used for regular alignment since they flip with locale direction. `"forceLeft"` and `"forceRight"` should only be used in special cases where locale direction should be ignored, such as tabular or numeric text. See the [alignment variant](https://gestalt.pinterest.systems/web/text#Alignment) for more details.
   */
  align?: 'start' | 'end' | 'center' | 'forceLeft' | 'forceRight';
  /**
   * The text content to be displayed.
   */
  children?: ReactNode;
  /**
   * The color of the text content. See the [colors variant](https://gestalt.pinterest.systems/web/text#Colors) for more details.
   */
  color?:
    | 'default'
    | 'disabled'
    | 'subtle'
    | 'success'
    | 'error'
    | 'warning'
    | 'shopping'
    | 'link'
    | 'inverse'
    | 'light'
    | 'dark';
  /**
   * Indicates how the text should flow with the surrounding content. See the [block vs inline variant](https://gestalt.pinterest.systems/web/text#Block-vs.-inline) for more details.
   */
  inline?: boolean;
  /**
   * Indicates if the text should be displayed in italic. See the [styles variant](https://gestalt.pinterest.systems/web/text#Styles) for more details.
   */
  italic?: boolean;
  /**
   * Visually truncate the text to the specified number of lines. This also adds the `title` attribute if `children` is a string, which displays the full text on hover in most browsers. See the [overflow and truncation variant](https://gestalt.pinterest.systems/web/text#Overflow-and-truncation) for more details.
   */
  lineClamp?: number;
  /**
   * Indicates how the text content should handle overflowing its container. See the [overflow and truncation variant](https://gestalt.pinterest.systems/web/text#Overflow-and-truncation) for more details.
   */
  overflow?: Overflow;
  /**
   * Ref that is forwarded to the underlying element. See the [ref variant](https://gestalt.pinterest.systems/web/text#Refs) for more details.
   */
  ref?: HTMLDivElement | HTMLSpanElement;
  /**
   * The sizes are based on our [font-size design tokens](https://gestalt.pinterest.systems/foundations/design_tokens/overview#Font-size). See the [sizes variant](https://gestalt.pinterest.systems/web/text#Sizes) for more details.
   */
  size?: Size;
  /**
   * This populates the `title` attribute of the element, which is visible on hover in most browsers. This is useful when truncating the text with `lineClamp` when `children` is a `React.Node`. See the [Title variant](https://gestalt.pinterest.systems/web/text#Title) for more details.
   */
  title?: string;
  /**
   * Indicates if the text content should be underlined. See the [styles variant](https://gestalt.pinterest.systems/web/text#Styles) for more details.
   */
  underline?: boolean;
  /**
   * Indicates the font weight for the text content. See the [styles variant](https://gestalt.pinterest.systems/web/text#Styles) for more details.
   */
  weight?: 'bold' | 'normal';
};

/**
 * [Text](https://gestalt.pinterest.systems/web/text) component is used for all non-heading text on all surfaces, whether inside of UI components or in long-form paragraph text.
 *
 * ![Text light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Text.spec.ts-snapshots/Text-chromium-darwin.png)
 * ![Text dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Text-dark.spec.ts-snapshots/Text-dark-chromium-darwin.png)
 */
const TextWithForwardRef = forwardRef<HTMLElement, Props>(function Text(
  {
    align = 'start',
    children,
    color = 'default',
    inline = false,
    italic = false,
    lineClamp,
    overflow = 'breakWord',
    size = '300',
    title,
    underline = false,
    weight = 'normal',
  }: Props,
  ref,
): ReactElement {
  const colorClass = semanticColors.includes(color) && styles[color];

  const getWordBreakStyle = (): string | undefined => {
    if (overflow === 'breakAll') {
      return typography.breakAll;
    }

    // default to breakWord if lineClamp is set
    if (overflow === 'breakWord' || isNotNullish(lineClamp)) {
      return typography.breakWord;
    }

    return undefined;
  };

  const cs = cx(
    styles.Text,
    typography[`fontSize${size}`],
    color && colorClass,
    align === 'center' && typography.alignCenter,
    // @ts-expect-error - TS2367 - This condition will always return 'false' since the types '"center" | "start" | "end" | "forceLeft" | "forceRight"' and '"justify"' have no overlap.
    align === 'justify' && typography.alignJustify,
    align === 'start' && typography.alignStart,
    align === 'end' && typography.alignEnd,
    align === 'forceLeft' && typography.alignForceLeft,
    align === 'forceRight' && typography.alignForceRight,
    getWordBreakStyle(),
    overflow === 'noWrap' && typography.noWrap,
    italic && typography.fontStyleItalic,
    underline && typography.underline,
    weight === 'bold' && typography.fontWeightSemiBold,
    weight === 'normal' && typography.fontWeightNormal,
    isNotNullish(lineClamp) && typography.lineClamp,
  );

  const Tag: As = inline ? 'span' : 'div';

  return (
    <Tag
      className={cs}
      title={
        title ?? (isNotNullish(lineClamp) && typeof children === 'string' ? children : undefined)
      }
      {...(lineClamp ? { style: { WebkitLineClamp: lineClamp } } : {})}
      // @ts-expect-error - TS2322 - Type 'ForwardedRef<HTMLElement>' is not assignable to type 'LegacyRef<HTMLDivElement> | undefined'.
      ref={ref}
    >
      {children}
    </Tag>
  );
});

TextWithForwardRef.displayName = 'Text';

export default TextWithForwardRef;
