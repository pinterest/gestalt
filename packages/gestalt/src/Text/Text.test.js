/* eslint-env jest */
import React from 'react';
import { create } from 'react-test-renderer';
import Text from './Text';

test('<Text>', () => {
  const tree = create(<Text />).toJSON();
  expect(tree).toMatchSnapshot();
});
