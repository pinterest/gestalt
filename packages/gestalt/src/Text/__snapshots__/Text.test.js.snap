// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text> 1`] = `
<div
  className="Text fontSize3 darkGray leadingShort alignLeft breakWord fontStyleNormal fontWeightNormal"
/>
`;
